src/transport/PeerMessageCounter.h

```diff
@@ -0,0 +1,189 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP message counters of remote nodes.
+ *
+ */
+#pragma once
+
+#include <array>
+#include <bitset>
+
+#include <support/Span.h>
+
+namespace chip {
+namespace Transport {
+
+class PeerMessageCounter
+{
+public:
+    static constexpr size_t kChallengeSize = 8;
+
+    PeerMessageCounter() : mStatus(Status::NotSynced) {}
+    ~PeerMessageCounter() { Reset(); }
+
+    void Reset()
+    {
+        switch (mStatus)
+        {
+        case Status::NotSynced:
+            break;
+        case Status::SyncInProcess:
+            mSyncInProcess.~SyncInProcess();
+            break;
+        case Status::Synced:
+            mSynced.~Synced();
+            break;
+        }
+        mStatus = Status::NotSynced;
+    }
+
+    bool IsSynchronizing() { return mStatus == Status::SyncInProcess; }
+    bool IsSynchronized() { return mStatus == Status::Synced; }
+
+    void SyncStarting(FixedByteSpan<kChallengeSize> challenge)
+    {
+        assert(mStatus == Status::NotSynced);
+        mStatus = Status::SyncInProcess;
+        new (&mSyncInProcess) SyncInProcess();
+        mSyncInProcess.mChallenge = challenge.ToArray();
+    }
+
+    void SyncFailed() { Reset(); }
+
+    CHIP_ERROR VerifyChallenge(uint32_t counter, FixedByteSpan<kChallengeSize> challenge)
+    {
+        if (mStatus != Status::SyncInProcess)
+        {
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+        if (mSyncInProcess.mChallenge != challenge.ToArray())
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+
+        mStatus = Status::Synced;
+        new (&mSynced) Synced();
```

Comment:
```suggestion
        mSyncInProcess.~SyncInProcess();
        new (&mSynced) Synced();
```
if we're going to do the whole ctor/dtor thing.

Reply:
It is very error prone to manually manipulate this constructor and destructor.

Can we use the Variant I have proposed in PR #6624 ?

Reply:
I agree it's pretty error-prone and would be happy to see something else...  A Variant-like thing would work, or encapsulating more of the logic inside the union itself (unions can have member functions) and coming up with some sort of "switch state" function that is implemented once.